File: java/advanced/LRUCache.java
package advanced;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * LRU (Least Recently Used) Cache — two implementations in one file:
 *
 * <ol>
 *   <li>{@link SimpleLRUCache} — leverages {@link LinkedHashMap} in access-order mode.
 *       One-liner {@code removeEldestEntry} handles eviction.</li>
 *   <li>{@link LRUCacheManual} — interview-style implementation using a doubly-linked
 *       list (DLL) + {@link java.util.HashMap}.  O(1) guaranteed for both get and put.</li>
 * </ol>
 *
 * <p>Time complexities (both versions):
 * <ul>
 *   <li>get – O(1)</li>
 *   <li>put – O(1)</li>
 * </ul>
 *
 * <p>Space complexity: O(capacity).
 */
public class LRUCache {

    // =========================================================================
    // 1. Simple version – LinkedHashMap
    // =========================================================================

    /**
     * Simple LRU Cache backed by a {@link LinkedHashMap} in access-order mode.
     *
     * <p>{@code LinkedHashMap(capacity, loadFactor, accessOrder=true)} keeps entries
     * in most-recently-accessed order.  Overriding {@link LinkedHashMap#removeEldestEntry}
     * enforces the capacity bound.
     */
    public static class SimpleLRUCache {

        private final int capacity;
        private final LinkedHashMap<Integer, Integer> map;

        /**
         * Creates a SimpleLRUCache with the given capacity.
         *
         * @param capacity maximum number of entries (>= 1)
         */
        public SimpleLRUCache(int capacity) {
            if (capacity < 1) throw new IllegalArgumentException("Capacity must be >= 1");
            this.capacity = capacity;
            // accessOrder=true: iterating visits LRU->MRU
            this.map = new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true) {
                @Override
                protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
                    return size() > capacity;
                }
            };
        }

        /**
         * Returns the value for {@code key}, or -1 if not present.
         * Marks the entry as most recently used.
         *
         * <p>Time: O(1) | Space: O(1).
         *
         * @param key the cache key
         * @return cached value or -1
         */
        public int get(int key) {
            return map.getOrDefault(key, -1);
        }

        /**
         * Inserts or updates the entry for {@code key} with {@code value}.
         * Evicts the least recently used entry when over capacity.
         *
         * <p>Time: O(1) | Space: O(1).
         *
         * @param key   cache key
         * @param value cache value
         */
        public void put(int key, int value) {
            map.put(key, value);
        }

        /** Returns the current number of entries. */
        public int size() { return map.size(); }

        /** Returns the cache capacity. */
        public int capacity() { return capacity; }

        @Override
        public String toString() {
            return "SimpleLRUCache" + map.toString() + " (LRU->MRU order)";
        }
    }

    // =========================================================================
    // 2. Manual version – DoublyLinkedList + HashMap (interview style)
    // =========================================================================

    /**
     * Manual LRU Cache using an explicit doubly-linked list and a HashMap.
     *
     * <p>Design:
     * <ul>
     *   <li>The DLL is kept in MRU order: {@code head.next} = most recently used,
     *       {@code tail.prev} = least recently used.</li>
     *   <li>Sentinel {@code head} and {@code tail} nodes simplify boundary conditions.</li>
     *   <li>The HashMap maps keys to their DLL nodes for O(1) access.</li>
     * </ul>
     *
     * <p>On {@code get}: move the accessed node to the front (MRU end).<br>
     * On {@code put}: if key exists, update value and move to front; otherwise
     * insert at front.  If over capacity, evict from the tail (LRU end).
     */
    public static class LRUCacheManual {

        // ----- Inner doubly-linked list node -----

        private static class DLLNode {
            int key, value;
            DLLNode prev, next;

            DLLNode(int key, int value) {
                this.key   = key;
                this.value = value;
            }
        }

        // ----- Fields -----

        private final int capacity;
        private final Map<Integer, DLLNode> map;
        private final DLLNode head; // sentinel MRU end
        private final DLLNode tail; // sentinel LRU end
        private int size;

        /**
         * Creates an LRUCacheManual with the given capacity.
         *
         * @param capacity maximum number of key-value pairs (>= 1)
         */
        public LRUCacheManual(int capacity) {
            if (capacity < 1) throw new IllegalArgumentException("Capacity must be >= 1");
            this.capacity = capacity;
            map  = new java.util.HashMap<>();
            head = new DLLNode(-1, -1); // dummy head (MRU side)
            tail = new DLLNode(-1, -1); // dummy tail (LRU side)
            head.next = tail;
            tail.prev = head;
        }

        // ----- Public API -----

        /**
         * Returns the value for {@code key}, or -1 if not present.
         * Marks the entry as most recently used.
         *
         * <p>Time: O(1) | Space: O(1).
         *
         * @param key the cache key
         * @return cached value or -1
         */
        public int get(int key) {
            DLLNode node = map.get(key);
            if (node == null) return -1;
            moveToFront(node);
            return node.value;
        }

        /**
         * Inserts or updates the entry for {@code key}.
         * Evicts the LRU entry if the cache would exceed capacity.
         *
         * <p>Time: O(1) | Space: O(1).
         *
         * @param key   cache key
         * @param value cache value
         */
        public void put(int key, int value) {
            DLLNode existing = map.get(key);
            if (existing != null) {
                existing.value = value;
                moveToFront(existing);
                return;
            }
            if (size == capacity) {
                // Evict LRU (node just before tail)
                DLLNode lru = tail.prev;
                removeNode(lru);
                map.remove(lru.key);
                size--;
            }
            DLLNode newNode = new DLLNode(key, value);
            addToFront(newNode);
            map.put(key, newNode);
            size++;
        }

        /** Returns the current number of entries. */
        public int size() { return size; }

        /** Returns the cache capacity. */
        public int capacity() { return capacity; }

        // ----- DLL helpers -----

        /** Detaches {@code node} from its current position. */
        private void removeNode(DLLNode node) {
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }

        /** Inserts {@code node} right after {@code head} (MRU position). */
        private void addToFront(DLLNode node) {
            node.next      = head.next;
            node.prev      = head;
            head.next.prev = node;
            head.next      = node;
        }

        /** Moves an existing {@code node} to the MRU position. */
        private void moveToFront(DLLNode node) {
            removeNode(node);
            addToFront(node);
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder("LRUCacheManual [MRU->LRU]: ");
            DLLNode cur = head.next;
            while (cur != tail) {
                sb.append("(").append(cur.key).append("=").append(cur.value).append(")");
                if (cur.next != tail) sb.append(" -> ");
                cur = cur.next;
            }
            sb.append("  | size=").append(size).append("/").append(capacity);
            return sb.toString();
        }
    }

    // =========================================================================
    // Main – demo for both implementations
    // =========================================================================

    public static void main(String[] args) {
        System.out.println("=== LRU Cache Demo ===\n");

        // ------------------------------------------------------------------
        // Simple version
        // ------------------------------------------------------------------
        System.out.println("--- SimpleLRUCache (capacity=3) ---");
        SimpleLRUCache simple = new SimpleLRUCache(3);

        simple.put(1, 10);
        simple.put(2, 20);
        simple.put(3, 30);
        System.out.println("After put(1,10) put(2,20) put(3,30): " + simple);

        System.out.println("get(1) = " + simple.get(1) + "  (accesses key 1 -> moves to MRU)");
        System.out.println("After get(1): " + simple);

        simple.put(4, 40); // evicts key 2 (LRU after get(1) made 1 MRU)
        System.out.println("After put(4,40) (evicts LRU): " + simple);
        System.out.println("get(2) = " + simple.get(2) + "  (expected -1, was evicted)");

        simple.put(5, 50);
        System.out.println("After put(5,50): " + simple);

        // ------------------------------------------------------------------
        // Manual version
        // ------------------------------------------------------------------
        System.out.println("\n--- LRUCacheManual (capacity=3) ---");
        LRUCacheManual lru = new LRUCacheManual(3);

        lru.put(1, 10);
        lru.put(2, 20);
        lru.put(3, 30);
        System.out.println("After put(1,10) put(2,20) put(3,30): " + lru);

        System.out.println("get(1) = " + lru.get(1) + "  (moves 1 to MRU)");
        System.out.println("After get(1): " + lru);

        lru.put(4, 40); // evicts key 2 (LRU)
        System.out.println("After put(4,40) (evicts 2): " + lru);
        System.out.println("get(2) = " + lru.get(2) + "  (expected -1)");
        System.out.println("get(3) = " + lru.get(3) + "  (expected 30)");
        System.out.println("After get(3): " + lru);

        lru.put(5, 50); // evicts LRU (currently key 1 or 4)
        System.out.println("After put(5,50): " + lru);

        lru.put(3, 300); // update existing – moves to front
        System.out.println("After put(3,300) [update]: " + lru);
        System.out.println("get(3) = " + lru.get(3) + "  (expected 300)");

        // ------------------------------------------------------------------
        // Classic LeetCode 146 sequence
        // ------------------------------------------------------------------
        System.out.println("\n--- LeetCode 146 sequence on LRUCacheManual ---");
        LRUCacheManual lc = new LRUCacheManual(2);
        lc.put(1, 1); System.out.println("put(1,1): " + lc);
        lc.put(2, 2); System.out.println("put(2,2): " + lc);
        System.out.println("get(1)=" + lc.get(1));      // 1
        lc.put(3, 3); System.out.println("put(3,3) evicts 2: " + lc);
        System.out.println("get(2)=" + lc.get(2));      // -1
        lc.put(4, 4); System.out.println("put(4,4) evicts 1: " + lc);
        System.out.println("get(1)=" + lc.get(1));      // -1
        System.out.println("get(3)=" + lc.get(3));      // 3
        System.out.println("get(4)=" + lc.get(4));      // 4
    }
}

File: python/advanced/lru_cache.py
"""
LRU Cache — Two Implementations
=================================
An LRU (Least Recently Used) cache evicts the entry that has not been
accessed for the longest time when the cache reaches capacity.

Implementation 1 — OrderedDict:
    Python's collections.OrderedDict maintains insertion order and supports
    O(1) move_to_end.  This makes the LRU cache trivially simple.

Implementation 2 — Doubly-Linked List + HashMap:
    The "canonical" interview implementation.  A doubly-linked list orders
    entries by recency (head = most recent, tail = least recent).
    A hash map provides O(1) node lookup by key.

Complexities (both implementations):
    - get(key):        O(1)
    - put(key, val):   O(1)
    - Space:           O(capacity)
"""

from __future__ import annotations
from collections import OrderedDict
from typing import Any, Dict, Optional


# ======================================================================
# Implementation 1: OrderedDict-based
# ======================================================================

class LRUCacheOD:
    """
    LRU Cache backed by collections.OrderedDict.

    OrderedDict.move_to_end(key) is O(1) and marks the key as most recently
    used.  popitem(last=False) removes the LRU entry in O(1).
    """

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("Capacity must be a positive integer.")
        self._capacity = capacity
        self._cache: OrderedDict = OrderedDict()

    def get(self, key: Any) -> Optional[Any]:
        """
        Return the value for *key*, or None if not present.
        Marks *key* as most recently used.

        Time:  O(1)
        """
        if key not in self._cache:
            return None
        # Move to the end (= most recently used position).
        self._cache.move_to_end(key)
        return self._cache[key]

    def put(self, key: Any, val: Any) -> None:
        """
        Insert or update *key* with *val*.
        Evicts the least recently used entry if at capacity.

        Time:  O(1)
        """
        if key in self._cache:
            self._cache.move_to_end(key)
        self._cache[key] = val
        if len(self._cache) > self._capacity:
            # popitem(last=False) removes the FIRST (oldest) item.
            self._cache.popitem(last=False)

    @property
    def size(self) -> int:
        return len(self._cache)

    def __repr__(self) -> str:
        # Show MRU → LRU order.
        items = list(reversed(self._cache.items()))
        return f"LRUCacheOD(cap={self._capacity}, items={items})"


# ======================================================================
# Implementation 2: Doubly-Linked List + HashMap
# ======================================================================

class _DLLNode:
    """A node in the doubly-linked list."""

    __slots__ = ("key", "val", "prev", "next")

    def __init__(self, key: Any = None, val: Any = None) -> None:
        self.key = key
        self.val = val
        self.prev: Optional[_DLLNode] = None
        self.next: Optional[_DLLNode] = None


class LRUCacheDLL:
    """
    LRU Cache using a Doubly-Linked List + HashMap.

    Layout:
        HEAD (dummy) ↔ [MRU node] ↔ ... ↔ [LRU node] ↔ TAIL (dummy)

    Sentinel HEAD and TAIL nodes eliminate edge-case checks for empty lists.

    Operations:
        get(key):
            1. Look up node in hashmap → O(1).
            2. Move node to just after HEAD → O(1).
        put(key, val):
            1. If key exists, update val and move to front.
            2. Else insert new node after HEAD.
            3. If over capacity, remove node just before TAIL (LRU).
    """

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("Capacity must be a positive integer.")
        self._capacity = capacity
        self._map: Dict[Any, _DLLNode] = {}

        # Sentinels — never removed, never stored in the hashmap.
        self._head = _DLLNode()   # most-recent end
        self._tail = _DLLNode()   # least-recent end
        self._head.next = self._tail
        self._tail.prev = self._head

    # ------------------------------------------------------------------
    # DLL helpers
    # ------------------------------------------------------------------

    def _remove(self, node: _DLLNode) -> None:
        """Unlink *node* from the DLL in O(1)."""
        node.prev.next = node.next
        node.next.prev = node.prev

    def _insert_after_head(self, node: _DLLNode) -> None:
        """Insert *node* right after HEAD (= most recently used) in O(1)."""
        node.prev = self._head
        node.next = self._head.next
        self._head.next.prev = node
        self._head.next = node

    def _move_to_front(self, node: _DLLNode) -> None:
        """Move existing *node* to the MRU position in O(1)."""
        self._remove(node)
        self._insert_after_head(node)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get(self, key: Any) -> Optional[Any]:
        """
        Return value for *key*, or None if absent.

        Time:  O(1)
        """
        if key not in self._map:
            return None
        node = self._map[key]
        self._move_to_front(node)
        return node.val

    def put(self, key: Any, val: Any) -> None:
        """
        Insert or update *key* / *val*.
        Evicts LRU entry if over capacity.

        Time:  O(1)
        """
        if key in self._map:
            node = self._map[key]
            node.val = val
            self._move_to_front(node)
        else:
            node = _DLLNode(key, val)
            self._map[key] = node
            self._insert_after_head(node)
            if len(self._map) > self._capacity:
                # Evict LRU: the node just before TAIL.
                lru = self._tail.prev
                self._remove(lru)
                del self._map[lru.key]

    @property
    def size(self) -> int:
        return len(self._map)

    def _to_list(self):
        """Return items as [(key, val), ...] in MRU → LRU order."""
        result = []
        cur = self._head.next
        while cur is not self._tail:
            result.append((cur.key, cur.val))
            cur = cur.next
        return result

    def __repr__(self) -> str:
        return f"LRUCacheDLL(cap={self._capacity}, items={self._to_list()})"


# ----------------------------------------------------------------------
# Demo
# ----------------------------------------------------------------------

if __name__ == "__main__":
    def run_demo(cache, label: str) -> None:
        print(f"=== {label} (capacity=3) ===")
        cache.put(1, "one")
        cache.put(2, "two")
        cache.put(3, "three")
        print("After put(1), put(2), put(3):", repr(cache))

        print("get(1):", cache.get(1))        # "one"  — 1 becomes MRU
        cache.put(4, "four")                  # evicts 2 (LRU)
        print("After get(1) + put(4):", repr(cache))

        print("get(2):", cache.get(2))        # None   — 2 was evicted
        print("get(3):", cache.get(3))        # "three"
        cache.put(5, "five")                  # evicts 4 (LRU)
        print("After put(5):", repr(cache))
        print()

    run_demo(LRUCacheOD(3), "LRUCache (OrderedDict)")
    run_demo(LRUCacheDLL(3), "LRUCache (DLL + HashMap)")

    # Stress-test equivalence.
    print("=== Equivalence stress test ===")
    od = LRUCacheOD(4)
    dll = LRUCacheDLL(4)
    ops = [
        ("put", "a", 1), ("put", "b", 2), ("put", "c", 3), ("put", "d", 4),
        ("get", "b", None), ("put", "e", 5),
        ("get", "a", None), ("get", "c", None),
        ("put", "f", 6), ("get", "d", None), ("get", "e", None),
    ]
    for op in ops:
        if op[0] == "put":
            od.put(op[1], op[2])
            dll.put(op[1], op[2])
        else:
            r1, r2 = od.get(op[1]), dll.get(op[1])
            assert r1 == r2, f"Mismatch on get({op[1]}): OD={r1}, DLL={r2}"
    print("All operations match between both implementations.")
